src/codecs/openexr.rs

```diff
@@ -0,0 +1,489 @@
+        // even though such a large allocation would probably cause an error much earlier
+        .ok_or_else(|| ImageError::Limits(LimitError::from_kind(LimitErrorKind::InsufficientMemory)))
+}
+
+
+// TODO this could be a generic function that works for any format
+/// Write an `Rgb32FImage` to the specified file path, replacing any existing file.
+// TODO progress? load rect?
+pub fn write_rgb_image_to_file(path: impl AsRef<Path>, image: &Rgb32FImage) -> ImageResult<()> {
+    write_rgb_image(BufWriter::new(std::fs::File::create(path)?), image)
+}
+
+// TODO this could be a generic function that works for any format
+/// Write an `Rgb32FImage`.
+/// Assumes the writer is buffered. In most cases,
+/// you should wrap your writer in a `BufWriter` for best performance.
+// TODO progress? load rect?
+pub fn write_rgb_image(write: impl Write/* + Seek*/, image: &Rgb32FImage) -> ImageResult<()> {
+    write_buffer(
+        write,
+        bytemuck::cast_slice(image.as_raw().as_slice()),
+        image.width(), image.height(),
+        ColorType::Rgb32F
+    )
+}
+
+// TODO this could be a generic function that works for any format
+/// Write an `Rgba32FImage` to the specified file path, replacing any existing file.
+// TODO progress? load rect?
+pub fn write_rgba_image_to_file(path: impl AsRef<Path>, image: &Rgba32FImage) -> ImageResult<()> {
+    write_rgba_image(BufWriter::new(std::fs::File::create(path)?), image)
+}
+
+// TODO this could be a generic function that works for any format
+/// Write an `Rgba32FImage`.
+/// Assumes the writer is buffered. In most cases,
+/// you should wrap your writer in a `BufWriter` for best performance.
+// TODO progress? load rect?
+pub fn write_rgba_image(write: impl Write/* + Seek*/, image: &Rgba32FImage) -> ImageResult<()> {
+    write_buffer(
+        write,
+        bytemuck::cast_slice(image.as_raw().as_slice()),
+        image.width(), image.height(),
+        ColorType::Rgba32F
+    )
+}
+
+
+/// An OpenEXR decoder. Immediately reads the meta data from the file.
+#[derive(Debug)]
+pub struct ExrDecoder<R> {
+    exr_reader: exr::block::reader::Reader<R>,
+
+    // select a header that is rgb and not deep
+    header_index: usize,
+
+    // decode either rgb or rgba.
+    // can be specified to include or discard alpha channels.
+    alpha_preference: Option<bool>,
+
+    alpha_present_in_file: bool,
+}
+
+
+impl<R: BufRead + Seek> ExrDecoder<R> {
+
+    /// Create a decoder. Consumes the first few bytes of the source to extract image dimensions.
+    /// Assumes the reader is buffered. In most cases,
+    /// you should wrap your reader in a `BufReader` for best performance.
+    /// If alpha preference is specified, an alpha channel will
+    /// always be present or always be not present in the returned image.
+    /// If alpha preference is none, the alpha channel will only be returned if it is found in the file.
+    pub fn read(source: R, alpha_preference: Option<bool>) -> ImageResult<Self> {
```

Comment:
Could you rename this `ExrDecoder::with_alpha_preference` and add a `ExrDecoder::new` method which assumes alpha_preference=None? That would be more consistent with other decoders

Reply:
Honestly, at some point we should really make the `new` method part of the ImageDecoder trait but there's some complications involved with that.

Reply:
> we should really make the new method part of the ImageDecoder

I thought so too some time before haha :)

Reply:
The embedded hal crates very much avoid having a constructor as part of the trait and I think this style is, for the most part, pretty clean. An alternative would be a method to 'suggest' a color type to an `ImageDecoder` so that you can modify the decoding later. Afterall, the choice need not be final after having read the header of an image.